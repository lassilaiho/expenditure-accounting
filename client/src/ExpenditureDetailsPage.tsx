import {
  AppBar,
  CircularProgress,
  Container,
  IconButton,
  Paper,
  Toolbar,
  Typography
} from '@material-ui/core';
import ArrowBackIcon from '@material-ui/icons/ArrowBack';
import { observer } from 'mobx-react';
import React from 'react';
import { Redirect, useHistory, useLocation } from 'react-router-dom';

import ExpenditureByTags from './ExpenditureByTags';
import { useProducts } from './product';
import { usePurchases } from './purchases';
import { useTags } from './tags';
import { DateRange } from './util';

export interface ExpenditureDetailsPageProps { }

const ExpenditureDetailsPage: React.FC<ExpenditureDetailsPageProps> = observer(props => {
  const location = useLocation();
  const history = useHistory();
  const purchaseStore = usePurchases();
  const productStore = useProducts();
  const tagStore = useTags();

  if (purchaseStore.dataState === 'not-started') {
    purchaseStore.getPurchases();
  }
  if (tagStore.dataState === 'not-started') {
    productStore.fetchTags();
  }

  const dateScope = parseDateScope(location.pathname);
  if (!dateScope) {
    return <Redirect to='/' />;
  }
  const isLoading = purchaseStore.dataState === 'loading'
    || tagStore.dataState === 'loading';

  return <>
    <AppBar position='sticky'>
      <Toolbar>
        <IconButton color='inherit' onClick={() => history.goBack()}>
          <ArrowBackIcon />
        </IconButton>
        <Typography variant='h6'>
          Expenditure on {dateScopeToString(dateScope)}
        </Typography>
      </Toolbar>
    </AppBar>
    <Container fixed>
      <Paper>
        {isLoading
          ? <CircularProgress color='secondary' />
          : <ExpenditureByTags
            purchases={purchaseStore.purchases}
            dateRange={dateScopeToRange(dateScope)} />}
      </Paper>
    </Container>
  </>;
});

type DateScope =
  | { type: 'day', day: number, month: number, year: number }
  | { type: 'month', month: number, year: number };

function dateScopeToString(s: DateScope) {
  switch (s.type) {
    case 'day': return `${s.day}.${s.month}.${s.year}`;
    case 'month': return `${s.month}/${s.year}`;
  }
}

function dateScopeToRange(s: DateScope) {
  switch (s.type) {
    case 'day':
      const date = new Date(s.year, s.month - 1, s.day);
      return new DateRange(date, date);
    case 'month':
      const from = new Date(s.year, s.month - 1, 1);
      const to = new Date(s.year, s.month, 0);
      return new DateRange(from, to);
  }
}

function parseDateScope(url: string): DateScope | null {
  const match = matchMonthAndYear(url);
  if (match) {
    return match;
  }
  return matchDate(url);
}

function matchDate(url: string): DateScope | null {
  const match = /^\/expenditure\/(\d+)\.(\d+)\.(\d+)$/.exec(url);
  if (!match) {
    return null;
  }
  const day = parseInt(match[1], 10);
  if (isNaN(day) || day < 1 || day > 31) {
    return null;
  }
  const month = parseInt(match[2], 10);
  if (isNaN(month) || month < 1 || month > 12) {
    return null;
  }
  const year = parseInt(match[3], 10);
  if (isNaN(year)) {
    return null;
  }
  return { type: 'day', day, month, year };
}

function matchMonthAndYear(url: string): DateScope | null {
  const match = /^\/expenditure\/(\d+)\/(\d+)$/.exec(url);
  if (!match) {
    return null;
  }
  const month = parseInt(match[1], 10);
  if (isNaN(month) || month < 1 || month > 12) {
    return null;
  }
  const year = parseInt(match[2], 10);
  if (isNaN(year)) {
    return null;
  }
  return { type: 'month', month, year };
}

export default ExpenditureDetailsPage;
