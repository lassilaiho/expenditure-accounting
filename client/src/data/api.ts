import moment from 'moment';

import * as jsonConv from './jsonConvert';

export type LoadState =
  | 'not-started'
  | 'loading'
  | 'finished'
  | 'failed';

export class SessionToken {
  public constructor(
    public readonly token: string,
    public readonly expiryTime: moment.Moment,
  ) { }

  public static fromJson(json: any): SessionToken {
    jsonConv.toObject(json);
    return new SessionToken(
      jsonConv.toString(json.token),
      jsonConv.toMomentUtc(json.expiryTime),
    );
  }

  public get isValid() {
    return this.expiryTime.isAfter(moment.utc());
  }

  public toJSON(): unknown {
    return {
      token: this.token,
      expiryTime: this.expiryTime.format(),
    };
  }
}

export default class Api {
  public baseUrl = '';

  public defaultHeaders: Record<string, string> = {
    'Authorization': ''
  };

  private _sessionToken: SessionToken | null = null;
  public get sessionToken() { return this._sessionToken; }
  public set sessionToken(token: SessionToken | null) {
    this._sessionToken = token;
    this.defaultHeaders['Authorization'] =
      token === null ? '' : 'Basic ' + btoa(token.token);
  }

  public constructor(baseUrl: string) {
    this.baseUrl = baseUrl;
  }

  public async get(url: string) {
    return fetch(this.baseUrl + url, {
      mode: 'cors',
      headers: { ...this.defaultHeaders },
    });
  }

  public async postJson(url: string, body: any) {
    return fetch(this.baseUrl + url, {
      mode: 'cors',
      method: 'POST',
      headers: {
        ...this.defaultHeaders,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    });
  }

  public async patchJson(url: string, body: any) {
    return fetch(this.baseUrl + url, {
      mode: 'cors',
      method: 'PATCH',
      headers: {
        ...this.defaultHeaders,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    });
  }

  public async delete(url: string) {
    return fetch(this.baseUrl + url, {
      mode: 'cors',
      method: 'DELETE',
      headers: { ...this.defaultHeaders },
    });
  }
}
