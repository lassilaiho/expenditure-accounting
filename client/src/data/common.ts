import { createAction } from '@reduxjs/toolkit';

import { Product } from './products';
import { Purchase } from './purchases';
import { Tag } from './tags';

export type Collection<T> = {
  byId: { [id: number]: T };
  all: number[];
};

export function emptyCollection<T>(): Collection<T> {
  return { byId: {}, all: [] };
}

export const id = <T>(x: T) => x;

export const clearRemoteData = createAction('clearRemoteData');
export type RemoteDataSet = [Purchase, Product, Tag[]][];
export const setRemoteData = createAction<RemoteDataSet>('setRemoteData');
