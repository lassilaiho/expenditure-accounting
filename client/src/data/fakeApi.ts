import Api from './api';

export type FakeResponse = (
  method: string,
  url: string,
  body?: any,
) => Promise<Response>;

export class FakeApi extends Api {
  public constructor(private fakeResponse: FakeResponse) {
    super('');
  }

  public async get(url: string) {
    return this.fakeResponse('get', url);
  }

  public async postJson(url: string, body: any) {
    return this.fakeResponse('post', url, body);
  }

  public async patchJson(url: string, body: any) {
    return this.fakeResponse('patch', url, body);
  }

  public async delete(url: string) {
    return this.fakeResponse('delete', url);
  }
}

function jsonResponse(body: any) {
  return new Response(JSON.stringify(body), {
    headers: {
      'Content-Type': 'application/json',
    },
    status: 200,
  });
}

const tags = [
  {
    id: 1,
    name: 'tag 1',
  },
  {
    id: 2,
    name: 'tag 2',
  },
  {
    id: 3,
    name: 'tag 3',
  },
];
const products = [
  {
    id: 1,
    name: 'product 1',
  },
  {
    id: 2,
    name: 'product 2',
  },
];
const purchases = [
  {
    id: 1,
    product: products[0],
    date: '2020-03-14',
    quantity: '1.0',
    price: '2.0',
    tags: [tags[0], tags[2]],
  },
  {
    id: 2,
    product: products[0],
    date: '2020-01-14',
    quantity: '10',
    price: '13.2',
    tags: [tags[2]],
  },
  {
    id: 3,
    product: products[1],
    date: '2020-03-14',
    quantity: '0.8',
    price: '0.2',
    tags: [tags[0], tags[1], tags[2]],
  },
];

export const fakeData = { tags, products, purchases };

export const fakeApi = new FakeApi(async (method, url, body) => {
  switch (method + url) {
    case 'get/purchases':
      return jsonResponse({ purchases });
  }
  return new Response(null, { status: 404 });
});
