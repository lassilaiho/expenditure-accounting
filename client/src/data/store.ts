import Big from 'big.js';
import {
  action,
  computed,
  makeObservable,
  observable,
  runInAction,
} from 'mobx';
import moment from 'moment';
import React, { useContext } from 'react';

import { ensureOk } from '../util';
import Api, { LoadState } from './api';
import * as jsonConv from './jsonConvert';

export class Tag {
  public id: number;
  public name: string;

  public constructor(id: number, name: string) {
    makeObservable(this, {
      id: observable,
      name: observable,
    });
    this.id = id;
    this.name = name;
  }

  public static fromJson(json: any) {
    jsonConv.toObject(json);
    return new Tag(jsonConv.toNumber(json.id), jsonConv.toString(json.name));
  }

  public lowerCaseMatch(s: string) {
    return this.name.toLocaleLowerCase().includes(s);
  }
}

export class Purchase {
  public id: number;
  public product: Product;
  public date: moment.Moment;
  public quantity: Big;
  public price: Big;
  public get totalPrice() {
    return this.quantity.mul(this.price);
  }
  public tags: Tag[];
  public get tagsSortedByName() {
    return this.tags.slice().sort((a, b) => {
      const aName = a.name.toLocaleLowerCase();
      const bName = b.name.toLocaleLowerCase();
      if (aName < bName) {
        return -1;
      }
      if (aName > bName) {
        return 1;
      }
      return 0;
    });
  }

  public constructor(
    id: number,
    product: Product,
    date: moment.Moment,
    quantity: Big,
    price: Big,
    tags: Tag[],
  ) {
    makeObservable(this, {
      id: observable,
      product: observable,
      date: observable.ref,
      quantity: observable.ref,
      price: observable.ref,
      totalPrice: computed,
      tags: observable,
      tagsSortedByName: computed,
    });
    this.id = id;
    this.product = product;
    this.date = date;
    this.quantity = quantity;
    this.price = price;
    this.tags = tags;
  }

  public static fromJson(json: any) {
    jsonConv.toObject(json);
    return new Purchase(
      jsonConv.toNumber(json.id),
      Product.fromJson(json.product),
      jsonConv.toMomentUtc(json.date),
      jsonConv.toBig(json.quantity),
      jsonConv.toBig(json.price),
      jsonConv.toArray(Tag.fromJson, json.tags),
    );
  }

  public lowerCaseMatch(s: string) {
    return (
      this.product.lowerCaseMatch(s) || this.tags.some(t => t.lowerCaseMatch(s))
    );
  }
}

export class Product {
  public id: number;
  public name: string;

  public constructor(id: number, name: string) {
    makeObservable(this, {
      id: observable,
      name: observable,
    });
    this.id = id;
    this.name = name;
  }

  public static fromJson(json: any) {
    jsonConv.toObject(json);
    return new Product(
      jsonConv.toNumber(json.id),
      jsonConv.toString(json.name),
    );
  }

  public lowerCaseMatch(s: string) {
    return this.name.toLocaleLowerCase().includes(s);
  }
}

export class Store {
  public dataState: LoadState = 'not-started';
  public purchases: Purchase[] = [];
  public get purchasesById() {
    const map = new Map<number, Purchase>();
    for (const p of this.purchases) {
      map.set(p.id, p);
    }
    return map;
  }
  public tagsById = new Map<number, Tag>();
  public get tags() {
    const tags: Tag[] = [];
    for (const tag of this.tagsById.values()) {
      tags.push(tag);
    }
    return tags;
  }
  private get tagsByName() {
    const map = new Map<string, Tag>();
    for (const tag of this.tagsById.values()) {
      map.set(tag.name.toLowerCase(), tag);
    }
    return map;
  }
  public productsById = new Map<number, Product>();
  public get products() {
    const products: Product[] = [];
    for (const product of this.productsById.values()) {
      products.push(product);
    }
    return products;
  }
  private get productsByName() {
    const map = new Map<string, Product>();
    for (const product of this.productsById.values()) {
      map.set(product.name.toLowerCase(), product);
    }
    return map;
  }

  public constructor(private api: Api) {
    makeObservable<
      Store,
      | 'tagsByName'
      | 'productsByName'
      | 'parsePurchaseJson'
      | 'parsePurchasesJson'
    >(this, {
      dataState: observable,
      purchases: observable,
      purchasesById: computed,
      tagsById: observable,
      tags: computed,
      tagsByName: computed,
      productsById: observable,
      products: computed,
      productsByName: computed,
      addTags: action,
      reloadData: action,
      parsePurchasesJson: action,
      parsePurchaseJson: action,
      restorePurchase: action,
    });
  }

  public async updatePurchase(id: number) {
    const purchase = this.purchasesById.get(id);
    if (purchase) {
      ensureOk(
        await this.api.patchJson(`/purchases/${purchase.id}`, {
          product: purchase.product.id,
          date: purchase.date.format(),
          price: purchase.price,
          quantity: purchase.quantity,
          tags: purchase.tags.map(t => t.id),
        }),
      );
    } else {
      throw new Error(`No purchase with id: ${id}`);
    }
  }

  public async addTags(names: string[]) {
    const newTags: string[] = [];
    const result: Tag[] = [];
    for (const name of names) {
      const tag = this.getTagByName(name);
      if (tag === undefined) {
        newTags.push(name);
      } else {
        result.push(tag);
      }
    }
    if (newTags.length > 0) {
      const resp = await this.api.postJson('/tags', { tags: newTags });
      ensureOk(resp);
      const respJson = await resp.json();
      runInAction(() => {
        for (const tagJson of respJson.tags) {
          const tag = Tag.fromJson(tagJson);
          this.tagsById.set(tag.id, tag);
          result.push(tag);
        }
      });
    }
    return result;
  }

  public async addProduct(name: string) {
    const product = this.getProductByName(name);
    if (product === undefined) {
      const resp = await this.api.postJson('/products', { name });
      ensureOk(resp);
      const product = Product.fromJson(await resp.json());
      runInAction(() => this.productsById.set(product.id, product));
      return product;
    }
    return product;
  }

  public async addPurchase(purchase: Purchase) {
    const resp = await this.api.postJson('/purchases', {
      product: purchase.product.id,
      date: purchase.date.format(),
      price: purchase.price,
      quantity: purchase.quantity,
      tags: purchase.tags.map(t => t.id),
    });
    ensureOk(resp);
    const respJson = jsonConv.toObject(await resp.json());
    const id = jsonConv.toNumber(respJson.id);
    runInAction(() => {
      purchase.id = id;
      for (let i = 0; i < this.purchases.length; i++) {
        if (purchase.date > this.purchases[i].date) {
          this.purchases.splice(i, 0, purchase);
          return;
        }
      }
      this.purchases.splice(this.purchases.length, 0, purchase);
    });
  }

  public async deletePurchase(id: number) {
    const i = this.purchases.findIndex(p => p.id === id);
    if (i < 0) {
      throw new Error(`Purchase with id ${id} does not exist`);
    }
    ensureOk(await this.api.delete(`/purchases/${id}`));
    runInAction(() => this.purchases.splice(i, 1));
  }

  public async restorePurchase(id: number) {
    const resp = await this.api.postJson(`/purchases/${id}/restore`, {});
    await ensureOk(resp);
    const purchase = this.parsePurchaseJson((await resp.json()).purchase);
    runInAction(() => {
      for (let i = 0; i < this.purchases.length; i++) {
        if (purchase.date > this.purchases[i].date) {
          this.purchases.splice(i, 0, purchase);
          return;
        }
      }
      this.purchases.splice(this.purchases.length, 0, purchase);
    });
  }

  public async getLatestPurchaseByProduct(productName: string) {
    productName = productName.toLocaleLowerCase();
    return (
      this.purchases.find(p => p.product.lowerCaseMatch(productName)) ?? null
    );
  }

  private getTagByName(name: string) {
    return this.tagsByName.get(name.toLowerCase());
  }

  private getProductByName(name: string) {
    return this.productsByName.get(name.toLowerCase());
  }

  public async reloadData() {
    try {
      this.dataState = 'loading';
      this.purchases.length = 0;
      this.tagsById.clear();
      this.productsById.clear();

      const resp = await this.api.get('/purchases');
      ensureOk(resp);

      const json = await resp.json();
      runInAction(() => {
        this.parsePurchasesJson(json);
        this.dataState = 'finished';
      });
    } catch (e) {
      runInAction(() => (this.dataState = 'failed'));
      throw e;
    }
  }

  private parsePurchasesJson(data: any) {
    const purchases = data.purchases;
    if (!Array.isArray(purchases)) {
      throw new Error('Invalid response json');
    }
    for (const obj of purchases) {
      this.purchases.push(this.parsePurchaseJson(obj));
    }
  }

  private parsePurchaseJson(data: any) {
    const purchase = Purchase.fromJson(data);
    const product = this.productsById.get(purchase.product.id);
    if (product) {
      purchase.product = product;
    } else {
      this.productsById.set(purchase.product.id, purchase.product);
    }
    for (let i = 0; i < purchase.tags.length; i++) {
      const currentTag = purchase.tags[i];
      const existingTag = this.tagsById.get(currentTag.id);
      if (existingTag) {
        purchase.tags[i] = existingTag;
      } else {
        this.tagsById.set(currentTag.id, currentTag);
      }
    }
    return purchase;
  }
}

export const StoreContext = React.createContext<Store | null>(null);

export function useStore() {
  const store = useContext(StoreContext);
  if (store === null) {
    throw new Error('Store must not be null');
  }
  if (store.dataState === 'not-started') {
    store.reloadData();
  }
  return store;
}
