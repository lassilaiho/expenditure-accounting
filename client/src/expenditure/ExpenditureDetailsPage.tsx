import { Paper } from '@material-ui/core';
import { observer } from 'mobx-react';
import moment from 'moment';
import React from 'react';
import { Redirect, useLocation } from 'react-router-dom';

import CenteredLoader from '../common/CenteredLoader';
import BackButton from '../common/BackButton';
import { useStore } from '../data/store';
import Scaffold from '../common/Scaffold';
import { DateRange } from '../util';
import ExpenditureByTags from './ExpenditureByTags';

const ExpenditureDetailsPage: React.FC = observer(() => {
  const location = useLocation();
  const store = useStore();

  const dateScope = parseDateScope(location.pathname);
  if (!dateScope) {
    return <Redirect to='/' />;
  }

  return (
    <Scaffold
      nav={<BackButton />}
      title={'Expenditure on ' + dateScopeToString(dateScope)}
      content={
        <Paper>
          {store.dataState === 'loading' ? (
            <CenteredLoader />
          ) : (
            <ExpenditureByTags
              purchases={store.purchases}
              dateRange={dateScopeToRange(dateScope)}
            />
          )}
        </Paper>
      }
    />
  );
});

type DateScope =
  | { type: 'day'; day: number; month: number; year: number }
  | { type: 'month'; month: number; year: number };

function dateScopeToString(s: DateScope) {
  switch (s.type) {
    case 'day':
      return `${s.day}.${s.month}.${s.year}`;
    case 'month':
      return `${s.month}/${s.year}`;
  }
}

function dateScopeToRange(s: DateScope) {
  switch (s.type) {
    case 'day': {
      const date = moment.utc([s.year, s.month - 1, s.day]);
      return new DateRange(date, date.clone().endOf('day'));
    }
    case 'month': {
      const date = moment.utc([s.year, s.month - 1]);
      return new DateRange(date, date.clone().endOf('month'));
    }
  }
}

function parseDateScope(url: string): DateScope | null {
  const match = matchMonthAndYear(url);
  if (match) {
    return match;
  }
  return matchDate(url);
}

function matchDate(url: string): DateScope | null {
  const match = /^\/expenditure\/(\d+)\.(\d+)\.(\d+)$/.exec(url);
  if (!match) {
    return null;
  }
  const day = parseInt(match[1], 10);
  if (isNaN(day) || day < 1 || day > 31) {
    return null;
  }
  const month = parseInt(match[2], 10);
  if (isNaN(month) || month < 1 || month > 12) {
    return null;
  }
  const year = parseInt(match[3], 10);
  if (isNaN(year)) {
    return null;
  }
  return { type: 'day', day, month, year };
}

function matchMonthAndYear(url: string): DateScope | null {
  const match = /^\/expenditure\/(\d+)\/(\d+)$/.exec(url);
  if (!match) {
    return null;
  }
  const month = parseInt(match[1], 10);
  if (isNaN(month) || month < 1 || month > 12) {
    return null;
  }
  const year = parseInt(match[2], 10);
  if (isNaN(year)) {
    return null;
  }
  return { type: 'month', month, year };
}

export default ExpenditureDetailsPage;
