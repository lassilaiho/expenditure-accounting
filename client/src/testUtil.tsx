import MomentUtil from '@date-io/moment';
import { MuiPickersUtilsProvider } from '@material-ui/pickers';
import { render } from '@testing-library/react';
import React, { useRef } from 'react';
import { MemoryRouter } from 'react-router-dom';

import { fakeApi } from './data/fakeApi';
import { Store, StoreContext } from './data/store';

const TestProviders = (defaultStore?: Store, history?: string[]) => (
  props: any,
) => {
  const store = useRef<Store | null>(null);
  if (store.current === null) {
    if (defaultStore === undefined) {
      store.current = new Store(fakeApi);
    } else {
      store.current = defaultStore;
    }
  }
  return (
    <StoreContext.Provider value={store.current}>
      <MuiPickersUtilsProvider utils={MomentUtil}>
        <MemoryRouter initialEntries={history}>{props.children}</MemoryRouter>
      </MuiPickersUtilsProvider>
    </StoreContext.Provider>
  );
};

export interface CustomRenderOptions {
  store?: Store;
  history?: string[];
}

const customRender = (ui: any, options?: CustomRenderOptions) => {
  let store, history;
  if (options) {
    store = options.store;
    history = options.history;
  }
  return render(ui, { wrapper: TestProviders(store, history) });
};

export * from '@testing-library/react';

export { customRender as render };
