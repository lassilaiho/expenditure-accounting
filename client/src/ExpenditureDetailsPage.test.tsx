import React from 'react';

import { fakeApi } from 'data/fakeApi';
import { Store } from 'data/store';
import ExpenditureDetailsPage from './ExpenditureDetailsPage';
import { render } from './testUtil';

test('renders correctly with invalid scope', () => {
  const { asFragment } = render(
    <ExpenditureDetailsPage />
  );
  expect(asFragment()).toMatchSnapshot();
});

test('renders correctly with day scope with data', async () => {
  const store = new Store(fakeApi);
  await store.reloadData();
  const { asFragment } = render(
    <ExpenditureDetailsPage />,
    { store, history: ['/expenditure/14.1.2020'] },
  );
  expect(asFragment()).toMatchSnapshot();
});

test('renders correctly with day scope without data', async () => {
  const { asFragment } = render(
    <ExpenditureDetailsPage />,
    { history: ['/expenditure/15.1.2020'] },
  );
  expect(asFragment()).toMatchSnapshot();
});

test('renders correctly with month scope with data', async () => {
  const store = new Store(fakeApi);
  await store.reloadData();
  const { asFragment } = render(
    <ExpenditureDetailsPage />,
    { store, history: ['/expenditure/3/2020'] },
  );
  expect(asFragment()).toMatchSnapshot();
});

test('renders correctly with month scope without data', async () => {
  const { asFragment } = render(
    <ExpenditureDetailsPage />,
    { history: ['/expenditure/4/2020'] },
  );
  expect(asFragment()).toMatchSnapshot();
});
