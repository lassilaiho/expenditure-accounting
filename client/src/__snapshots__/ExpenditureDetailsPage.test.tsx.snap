// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with day scope with data 1`] = `
<DocumentFragment>
  <header
    class="MuiPaper-root MuiAppBar-root MuiAppBar-positionSticky MuiAppBar-colorPrimary MuiPaper-elevation4"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <h6
        class="MuiTypography-root MuiTypography-h6"
      >
        Expenditure on 14.1.2020
      </h6>
    </div>
  </header>
  <div
    class="MuiContainer-root MuiContainer-fixed MuiContainer-maxWidthLg"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
    >
      <ul
        class="MuiList-root MuiList-padding"
      >
        <li
          class="MuiListItem-root MuiListItem-gutters"
        >
          <div
            class="MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
            >
              <div
                class="MuiBox-root MuiBox-root-1"
              >
                <div
                  class="MuiBox-root MuiBox-root-2"
                >
                  tag 3
                </div>
                <div
                  class="MuiBox-root MuiBox-root-3"
                >
                  132,00 €
                </div>
              </div>
            </span>
          </div>
        </li>
        <li
          class="MuiListItem-root MuiListItem-gutters"
        >
          <div
            class="MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
            >
              <div
                class="MuiBox-root MuiBox-root-4"
              >
                <div
                  class="MuiBox-root MuiBox-root-5"
                >
                  untagged
                </div>
                <div
                  class="MuiBox-root MuiBox-root-6"
                >
                  0,00 €
                </div>
              </div>
            </span>
          </div>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders correctly with day scope without data 1`] = `
<DocumentFragment>
  <header
    class="MuiPaper-root MuiAppBar-root MuiAppBar-positionSticky MuiAppBar-colorPrimary MuiPaper-elevation4"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <h6
        class="MuiTypography-root MuiTypography-h6"
      >
        Expenditure on 15.1.2020
      </h6>
    </div>
  </header>
  <div
    class="MuiContainer-root MuiContainer-fixed MuiContainer-maxWidthLg"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiCircularProgress-root MuiCircularProgress-colorSecondary MuiCircularProgress-indeterminate"
        role="progressbar"
        style="width: 40px; height: 40px;"
      >
        <svg
          class="MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders correctly with invalid scope 1`] = `<DocumentFragment />`;

exports[`renders correctly with month scope with data 1`] = `
<DocumentFragment>
  <header
    class="MuiPaper-root MuiAppBar-root MuiAppBar-positionSticky MuiAppBar-colorPrimary MuiPaper-elevation4"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <h6
        class="MuiTypography-root MuiTypography-h6"
      >
        Expenditure on 3/2020
      </h6>
    </div>
  </header>
  <div
    class="MuiContainer-root MuiContainer-fixed MuiContainer-maxWidthLg"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
    >
      <ul
        class="MuiList-root MuiList-padding"
      >
        <li
          class="MuiListItem-root MuiListItem-gutters"
        >
          <div
            class="MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
            >
              <div
                class="MuiBox-root MuiBox-root-10"
              >
                <div
                  class="MuiBox-root MuiBox-root-11"
                >
                  tag 1
                </div>
                <div
                  class="MuiBox-root MuiBox-root-12"
                >
                  2,16 €
                </div>
              </div>
            </span>
          </div>
        </li>
        <li
          class="MuiListItem-root MuiListItem-gutters"
        >
          <div
            class="MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
            >
              <div
                class="MuiBox-root MuiBox-root-13"
              >
                <div
                  class="MuiBox-root MuiBox-root-14"
                >
                  tag 3
                </div>
                <div
                  class="MuiBox-root MuiBox-root-15"
                >
                  2,16 €
                </div>
              </div>
            </span>
          </div>
        </li>
        <li
          class="MuiListItem-root MuiListItem-gutters"
        >
          <div
            class="MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
            >
              <div
                class="MuiBox-root MuiBox-root-16"
              >
                <div
                  class="MuiBox-root MuiBox-root-17"
                >
                  tag 2
                </div>
                <div
                  class="MuiBox-root MuiBox-root-18"
                >
                  0,16 €
                </div>
              </div>
            </span>
          </div>
        </li>
        <li
          class="MuiListItem-root MuiListItem-gutters"
        >
          <div
            class="MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
            >
              <div
                class="MuiBox-root MuiBox-root-19"
              >
                <div
                  class="MuiBox-root MuiBox-root-20"
                >
                  untagged
                </div>
                <div
                  class="MuiBox-root MuiBox-root-21"
                >
                  0,00 €
                </div>
              </div>
            </span>
          </div>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders correctly with month scope without data 1`] = `
<DocumentFragment>
  <header
    class="MuiPaper-root MuiAppBar-root MuiAppBar-positionSticky MuiAppBar-colorPrimary MuiPaper-elevation4"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <h6
        class="MuiTypography-root MuiTypography-h6"
      >
        Expenditure on 4/2020
      </h6>
    </div>
  </header>
  <div
    class="MuiContainer-root MuiContainer-fixed MuiContainer-maxWidthLg"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiCircularProgress-root MuiCircularProgress-colorSecondary MuiCircularProgress-indeterminate"
        role="progressbar"
        style="width: 40px; height: 40px;"
      >
        <svg
          class="MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
